### PR TITLE
AMLS-4463

### PR DESCRIPTION
**Pull request by owen-opencast**
DES connectors perform retries on 5xx range HTTP errors, with exponential back-off configurable from application.conf

## Checklist

- [ ] Breaking change (fix or feature that would cause existing functionality to change).
- [ ] Build passing.
- [ ] Unit tests have full coverage.
- [x] Unit test approach and implementation has been reviewed with QA (testers).
- [x] Requires acceptance test run.
- [ ] Appropriate labels added.

### CODE REVIEW

File: app/utils/BackOffHelper.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import akka.actor.ActorSystem
+import akka.pattern.Patterns.after
+import config.AmlsConfig
+import exceptions.HttpStatusException
+import javax.inject._
+import play.api.Logger
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+
+trait BackOffHelper  {
```
Comment (ajchilts) [inline app/utils/BackOffHelper.scala]: Possibly rename this to something more descriptive like RertyHelper or APIHelper? It executes the function and therefore does more than handle an exponential back off.

**Comment by slavvco** [comment]
Please mark this pull request with appropriate labels, to make it clear on what stage is the development ;)

File: test/utils/BackOffHelperSpec.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import play.api.inject.bind
+import play.api.inject.guice.GuiceApplicationBuilder
+import play.api.test.Helpers._
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class BackOffHelperSpec extends PlaySpec with MockitoSugar with ScalaFutures with OneAppPerSuite {
+
+  override lazy val app = new GuiceApplicationBuilder()
```
Comment (wolfendale) [inline test/utils/BackOffHelperSpec.scala]: There's no need to instantiate an app here considering all of your set up is just to get an instance of a thing which usually doesn't exist in the application. Just create an instance yourself. This means that you can also remove `OneAppPerSuite` which will speed up the tests.

```scala
val backOffHelper = new BackOffHelper()
```

File: app/utils/BackOffHelper.scala (hunk 1 above)
Comment (samzaroughmrc) [reply app/utils/BackOffHelper.scala]: @ajchilts class is now renamed to ApiRetryHelper

File: app/connectors/DeregisterSubscriptionConnector.scala [hunk 3]
```diff
@@ -27,11 +27,23 @@ import play.api.http.Status.{INTERNAL_SERVER_ERROR, OK}
...
+                    (
+                    implicit ec: ExecutionContext,
+                    wr1: Writes[DeregisterSubscriptionRequest],
+                    wr2: Writes[DeregisterSubscriptionResponse],
+                    hc: HeaderCarrier,
+                    apiRetryHelper: ApiRetryHelper
+  ): Future[DeregisterSubscriptionResponse] = {
+    apiRetryHelper.doWithBackoff(() => deregistrationFunction(amlsRegistrationNumber, data))
+  }
+
```
[the remark this answers is not in the document]
Comment (samzaroughmrc) [reply app/connectors/DeregisterSubscriptionConnector.scala]: yes and perhaps no. but i guess better than the long line previously

File: app/controllers/AmendVariationController.scala [hunk 4]
```diff
@@ -90,6 +96,7 @@ trait AmendVariationController extends BaseController {
       implicit request =>
         val prefix = "[AmendVariationController][amend]"
         Logger.debug(s"$prefix - AmlsRegistrationNumber: $amlsRegistrationNumber")
+        implicit val as: ActorSystem = ActorSystem()
```
[the remark this answers is not in the document]
Comment (samzaroughmrc) [reply app/controllers/AmendVariationController.scala]: good call thanks!

**Pull request merged**